backend-ktor/src/main/kotlin/me/zwsmith/plugins/database/RecipeService.kt

```diff
@@ -0,0 +1,135 @@
+package me.zwsmith.plugins.database
+
+import kotlinx.coroutines.Dispatchers
+import me.zwsmith.plugins.Ingredient
+import me.zwsmith.plugins.QuantifiedIngredient
+import me.zwsmith.plugins.Recipe
+import me.zwsmith.plugins.recipes.QuantifiedIngredientInput
+import org.jetbrains.exposed.sql.Database
+import org.jetbrains.exposed.sql.SchemaUtils
+import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
+import org.jetbrains.exposed.sql.transactions.transaction
+
+class RecipeService(database: Database) {
+    init {
+        transaction(database) {
+            SchemaUtils.create(Recipes, Ingredients, QuantifiedIngredients)
+        }
+    }
+    
+    suspend fun createIngredient(name: String): Ingredient = dbQuery {
+        IngredientDao.new {
+            this.name = name
+        }.toIngredient()
+    }
+    
+    suspend fun findIngredientById(id: Int) = dbQuery {
+        IngredientDao.findById(id)?.toIngredient()
+    }
+    
+    suspend fun updateIngredient(updateData: Ingredient) = dbQuery {
+        IngredientDao.findById(updateData.id)?.apply {
+            name = updateData.name
+        }?.toIngredient()
+    }
+    
+    suspend fun deleteIngredient(id: Int) = dbQuery {
+        val ingredient = IngredientDao.findById(id)
+        if(ingredient == null) {
+            throw IllegalArgumentException("No Ingredient found with ID $id")
+        } else {
+            ingredient.delete()
+        }
+    }
+    
+    suspend fun createRecipe(
+        name: String,
+        description: String,
+        ingredientInputs: List<QuantifiedIngredientInput>
+    ): Recipe {
+        return dbQuery {
+            val recipe = RecipeDao.new {
+                this.name = name
+                this.description = description
+            }
+            
+            // Associate QuantifiedIngredients with the new Recipe
+            ingredientInputs.forEach { input ->
+                val ingredient = IngredientDao
+                    .find { Ingredients.name eq input.ingredient.name }
+                    .singleOrNull()
+                    ?: IngredientDao.new {
+                        this.name = input.ingredient.name
+                    }
+                QuantifiedIngredientDao.new {
+                    this.amount = input.amount
+                    this.unit = input.unit
+                    this.recipe = recipe
+                    this.ingredient = ingredient
+                }
+            }
+            
+            recipe.toRecipe()
+        }
+    }
+    
+    suspend fun findRecipeById(id: Int): Recipe = dbQuery {
+        RecipeDao.findById(id)
+            ?.toRecipe()
+            ?: throw IllegalArgumentException("Recipe with id $id does not exist")
```

Comment:
I don't think throwing an exception here is the right move - that seems aggressive. Returning null/optional seems more expressive and ergonomic

Reply:
Yea error handling in general has been something I've been unsure about. Like if this was and Android app with my usual unidirectional data flow approach I would just wrap everything in results but was thinking maybe that was over engineering/premature. You think I should update the service methods to expose `Results`?